benchmarks/torchbench/voirfile.py

```diff
@@ -1,11 +1,212 @@
+                )
+            elif model == "LearningToPaint":
+                pass
+            elif model == "dcgan":
+                (
+                    ov.probe(f"/torchbenchmark.models.{model}/Model/train(benchmark_pic as batch) > #endloop_i as step")
+                    .augment(batch_size=lambda batch: len(batch))
+                    .give()
+                )
+                loss = (
+                    ov.probe(f"/torchbenchmark.models.{model}/Model/train > errD as loss")
+                )
+            elif model == "detectron2_maskrcnn":
+                (
+                    ov.probe(f"/torchbenchmark.models.{model}/Model/train(data as batch) > #endloop_data as step")
+                    .augment(batch_size=lambda batch: len(batch))
+                    .give()
+                )
+                loss = (
+                    ov.probe(f"/torchbenchmark.models.{model}/Model/train > loss")
+                )
+            elif model in {"drq", "soft_actor_critic"}:
+                (
+                    ov.probe(f"/torchbenchmark.models.{model}/Model/train > #endloop_step as step")
+                    .give()
+                )
+                (
+                    ov.probe(f"/torchbenchmark.models.{model}/Model/train > reward")
+                    .slice(-1)["reward"]
+                    .map(float)
+                    .give("reward")
+                )
+            elif model == "fastNLP_Bert":
+                (
+                    ov.probe(f"/torchbenchmark.models.{model}/Model/train(batch_x as batch) > #endloop_batch_x as step")
+                    .augment(batch_size=lambda batch: len(batch))
+                    .give()
+                )
+                loss = (
+                    ov.probe(f"/torchbenchmark.models.{model}/Model/train > loss")
+                )
+            elif model in {"maml", "maml_omniglot"}:
+                (
+                    ov.probe(f"/torchbenchmark.models.{model}/Model/train(self) > #endloop__ as step")
+                    .augment(batch=lambda self: self.example_inputs[0])
+                    .augment(batch_size=lambda batch: len(batch))
+                    .give()
+                )
+            elif model in {"mobilenet_v2_quantized_qat", "resnet50_quantized_qat"}:
+                (
+                    ov.probe(f"/torchbenchmark.models.{model}/Model/train(self) > #endloop__ as step")
+                    .augment(batch=lambda self: self.example_inputs[0])
+                    .augment(batch_size=lambda batch: len(batch))
+                    .give()
+                )
+                @ov.probe(f"/torchbenchmark.models.{model}/Model/train > loss").ksubscribe
+                def reg(loss):
+                    (
+                        ov.probe("loss.__call__() as loss")
+                        .last()["loss"]
+                        .map(float)
+                        .give("loss")
+                    )
```

Comment:
This works but produces multiple loss outputs at the end of the run:
```
resnet50_quantized_qat.0 {'loss': 8.192030906677246}
resnet50_quantized_qat.0 {'loss': 8.006749153137207}
resnet50_quantized_qat.0 {'loss': 8.51313304901123}
resnet50_quantized_qat.0 {'loss': 8.55726432800293}
resnet50_quantized_qat.0 {'loss': 7.348790645599365}
resnet50_quantized_qat.0 {'loss': 7.04877233505249}
resnet50_quantized_qat.0 {'loss': 7.06369686126709}
resnet50_quantized_qat.0 {'loss': 6.985150337219238}
resnet50_quantized_qat.0 {'loss': 6.974846839904785}
resnet50_quantized_qat.0 {'loss': 7.002440929412842}
resnet50_quantized_qat.0 {'loss': 7.012561798095703}
resnet50_quantized_qat.0 {'loss': 6.938030242919922}
resnet50_quantized_qat.0 {'loss': 6.938986301422119}
```
This is because the loss gets created at every iteration (3+10). Any way we could only keep the last loss?

Reply:
I've also made some tests on the performance (GPU Time in ms) with and without probes:

* With no probes at all (`pass` a the beginning of `instrument_probe`):
```
444.420
444.489
445.219
445.754
444.923
445.255
445.240
445.230
445.349
444.539
``` 
* With only steps probs:
```python
            elif model in {"mobilenet_v2_quantized_qat", "resnet50_quantized_qat"}:
                (
                    ov.probe(f"/torchbenchmark.models.{model}/Model/train(self) > #endloop__ as step")
                    .augment(batch=lambda self: self.example_inputs[0])
                    .augment(batch_size=lambda batch: len(batch))
                    .give()
                )
                # @ov.probe(f"/torchbenchmark.models.{model}/Model/train > loss").ksubscribe
                # def reg(loss):
                #     (
                #         ov.probe("loss.__call__() as loss")
                #         .last()["loss"]
                #         .map(float)
                #         .give("loss")
                #     )
```
```
445.585
444.524
445.407
445.228
444.937
445.389
444.827
446.447
445.801
445.347
```
* With only steps and loss probs:
```python
            elif model in {"mobilenet_v2_quantized_qat", "resnet50_quantized_qat"}:
                (
                    ov.probe(f"/torchbenchmark.models.{model}/Model/train(self) > #endloop__ as step")
                    .augment(batch=lambda self: self.example_inputs[0])
                    .augment(batch_size=lambda batch: len(batch))
                    .give()
                )
                @ov.probe(f"/torchbenchmark.models.{model}/Model/train > loss").ksubscribe
                def reg(loss):
                    (
                        ov.probe("loss.__call__() as loss")
                        .last()["loss"]
                        .map(float)
                        .give("loss")
                    )
```
```
474.104
472.799
472.168
472.240
471.581
470.290
472.046
471.129
471.912
473.887
```

So there seam to be some performance drop from those probes. I currently suspects gpu sync but this could be investigated. It might not be as big of an issue in real training loops but this could also be investigated.

Reply:
Interesting, thanks for checking! GPU sync normally should happen with the step probe. I think the issue is that you're running `ov.probe` on every `loss`. I think that this might work better (use `first()` on the outer probe rather than `last()` on the inner one):

```python
                @ov.probe(f"/torchbenchmark.models.{model}/Model/train > loss").first().ksubscribe
                def reg(loss):
                    (
                        ov.probe("loss.__call__() as loss")
                        ["loss"]
                        .map(float)
                        .give("loss")
                    )
```

Reply:
Yes I also think that is the issue but I was more interested with the last lost than the first. Thinking about it now there that might not be better. The main issue seams to be that loss gets defined in the train func:

https://github.com/pytorch/benchmark/blob/ff7114655294aa3ba57127a260dbd1ef5190f610/torchbenchmark/models/resnet50_quantized_qat/__init__.py#L44-L52

so a new loss is created for every iteration (and every iteration contains 3 sub loop on which the loss is computed)

Reply:
The probe can be placed on the class.

```python
                @ov.probe(f"/torchbenchmark.models.{model}/Model/train > loss").first().ksubscribe
                def reg(loss):
                    loss_type = type(loss)
                    (
                        ov.probe("loss_type.__call__() as loss")
                        ["loss"]
                        .map(float)
                        .give("loss")
                    )
```

In practice that will instrument all calls to CrossEntropyLoss the first time it sees `loss`. It's a bit similar to what I do to instrument the loader.

Reply:
Unfortunately the last suggestion (`ov.probe("loss_type.__call__() as loss")`) doesn't seam to run:
```
ValueError: only one element tensors can be converted to Python scalars
```

But I'm not entirely sure how it differs from the previous suggestion (`ov.probe("loss.__call__() as loss")`) which gives the first 3 loss rather than the last loss:

```
resnet50_quantized_qat.0 {'loss': 7.063980579376221}
resnet50_quantized_qat.0 {'loss': 7.721994876861572}
resnet50_quantized_qat.0 {'loss': 8.192975997924805}
resnet50_quantized_qat.0 {'train_rate': 209.82415609226032}
resnet50_quantized_qat.0 {'train_rate': 211.878567321225}
resnet50_quantized_qat.0 {'train_rate': 211.91041223191704}
resnet50_quantized_qat.0 {'train_rate': 213.49795693130113}
resnet50_quantized_qat.0 {'train_rate': 213.78586813280126}
resnet50_quantized_qat.0 {'gpu_time': 449.8754425048828}
resnet50_quantized_qat.0 {'cpu_dispatch_time': 448.3342635}
resnet50_quantized_qat.0 {'cpu_walltime': 449.894494}
resnet50_quantized_qat.0 > GPU Time:            449.875 milliseconds
resnet50_quantized_qat.0 > CPU Dispatch Time:   448.334 milliseconds
resnet50_quantized_qat.0 > CPU Total Wall Time: 449.894 milliseconds
resnet50_quantized_qat.0 {'train_rate': 213.93679953373012}
resnet50_quantized_qat.0 End: 2022-04-11 11:42:09.295972
```

It's alright if we prefer to use the first loss rather than the last. I'm not sure anymore there is much more value in the last anyway. Thoughts?

Reply:
> ValueError: only one element tensors can be converted to Python scalars

That seems like `float` crapping out on a multi-element tensor, which is a bit strange.

> But I'm not entirely sure how it differs from the previous suggestion (ov.probe("loss.__call__() as loss")) which gives the first 3 loss rather than the last loss:

The difference is that the first will only trigger for a particular `self`, whereas the latter will trigger for all instances. Basically, `probe("x.y() as z")` when `x` is an object should work the same as `probe("type_of_x.y(self = <x>) as z")`.

> It's alright if we prefer to use the first loss rather than the last. I'm not sure anymore there is much more value in the last anyway. Thoughts?

We mostly want the loss to serve as a sanity check, for example to verify that there is no `NaN`, or that the training converges to a set known value given the seed that we have. Having the last loss is better because it lets us know that we never had any `NaN`, whereas with the first loss we could still have them in a subsequent training step.

We can give this a lesser priority for now, if there are still a few benchmarks to instrument.

Reply:
I made some tests with a regular resnet18

Current code:
```python
    # loss of TorchVisionModel
    @ov.probe("//run_one_step > func").kmap(model=lambda func: func.__self__).ksubscribe
    def reg_vision_loss(model):
        if hasattr(model, "loss_fn"):
            (
                ov.probe("model.loss_fn.__call__() as loss")
                .last()["loss"]
                .map(float)
                .give("loss")
            )
```
GPU Time is about 67ms

Without loss:
GPU Time is about 60 ms

With first loss:
```python
    # loss of TorchVisionModel
    @ov.probe("//run_one_step > func").kmap(model=lambda func: func.__self__).ksubscribe
    def reg_vision_loss(model):
        if hasattr(model, "loss_fn"):
            (
                ov.probe("model.loss_fn.__call__() as loss")
                .first()["loss"]
                .map(float)
                .give("loss")
            )
```
GPU Time is about 69ms

I'm surprise to see a bit lower performance with the first loss but that might only be caused by the variance (I only checked 5 runs this time)